### DISCUSSION TITLE
Proposed `approvers-*` teams for PR reviews

### DISCUSSION THREAD

**Discussion by themarwhal** [magma/magma#6218] [Ideas]
# Proposed `approvers-*` teams for PR reviews

This proposal outlines the GitHub teams necessary to fully migrate to GitHub team based PR reviews.

I will create separate proposal on managing member permissions and membership change-logs for these  `approvers-*` teams.

## GitHub team based PR Reviews

In order to streamline the PR review workflow, we are migrating to a GitHub team based approach. 
To make this change, the `CODEOWNERS` file will be modified so that each path points to a GitHub team, rather than a list of individuals.  

Please refer to the [slides](https://docs.google.com/presentation/d/1D8DBCczojIQIOyaXwg4VRPNHf1SIN34nc3x8B-fbzdI/edit?usp=sharing) I made on this for example workflows. 

## Proposed teams

| team name | members |
| ------------------- | -------------------- |
| approvers-ci | @tmdzk @mattymo @119Vik @rdefosse |
| approvers-feg | @mpgermano @emakeev @themarwhal @uri200 |
| approvers-nms | @karthiksubraveti @andreilee @HannaFar |
| approvers-openwrt | @emakeev @uri200 |
| approvers-orc8r | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti |
| approvers-xwf | @AyliD @aharonnovo @r-i-g |
| approvers-agw | @themarwhal @ardzoht @ssanadhya @ulaskozat @pshelar @uri200 @koolzz |
| approvers-agw-c | @themarwhal @ardzoht |
| approvers-agw-python | @ardzoht @pshelar |
| approvers-agw-mme | @ssanadhya @ulaskozat @lionelgo @ardzoht |
| approvers-agw-sessiond | @themarwhal @uri200 |
| approvers-agw-pipelined | @koolzz @pshelar |
| approvers-agw-integ-tests | @ssanadhya @ulaskozat @ardzoht |

Note: At the request of Hunter, I’ll be changing the team names of `orc8r-approvers` and `feg-approvers` to be prefixed by `approvers-` so that all the teams show up sorted next to each other. 

## Change-list per directory

The intention was to keep the code ownership as unchanged as possible. However, in order to keep the set of `approvers-*` teams manageable, I’ve had to expand the ownership for a few directories. The `changed` column in the table below indicates the affected directories. 

The two main changes are as follows:

1. In order to have a single `approvers-ci` team, I’ve extended ownership of some of the CI related directories to `@tmdzk @mattymo @119Vik @rdefosse`.
2. I’ve separated out lte/gateway ownership and lte/gateway/python ownership. Previously, @ardzoht and  @pshelar held ownership for all non-specific files in the lte/gateway directory which included configs. I’ve instead created an approvers-agw team that has all AGW members in order to remove some PR review load from @ardzoht and  @pshelar for the non-service files. 

| directory | team name | previous owners | new owners | changed |
| ------------------- | -------------------- | -------------- | --------- | --------- |
| .circleci/ | approvers-ci | @tmdzk | @tmdzk @mattymo @119Vik @rdefosse | Y | 
| circleci | approvers-ci | @tmdzk @mattymo @119Vik | @tmdzk @mattymo @119Vik @rdefosse | Y | 
| ci-scripts/ | approvers-ci | @rdefosse @tmdzk @mattymo @119Vik | @tmdzk @mattymo @119Vik @rdefosse | N | 
| */cloud/ | approvers-orc8r | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | N | 
| .golangci.yml | approvers-orc8r | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | N |
| orc8r/ | approvers-orc8r | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | @andreilee @mpgermano @emakeev @hcgatewood @karthiksubraveti | N |
| orc8r/tools/packer/ | approvers-ci | @tmdzk @mattymo @119Vik | @tmdzk @mattymo @119Vik @rdefosse | Y |
| orc8r/cloud/deploy/bare-metal/ | approvers-ci | @tmdzk @mattymo @119Vik | @tmdzk @mattymo @119Vik @rdefosse | Y |
| orc8r/cloud/deploy/bare-metal-ansible/ | approvers-ci | @tmdzk @mattymo @119Vik | @tmdzk @mattymo @119Vik @rdefosse | Y | 
| orc8r/gateway/c/ | approvers-agw-c | @themarwhal @ardzoht | @themarwhal @ardzoht | N |
| feg/ | approvers-feg | @mpgermano @emakeev @themarwhal @uri200 | @mpgermano @emakeev @themarwhal @uri200 | N |
| cwf/ | approvers-cwf | @themarwhal @mpgermano @uri200 @emakeev | @themarwhal @mpgermano @uri200 @emakeev | N |
| cwf/gateway/Vagrantfile | approvers-ci | @mattymo @119Vik @tmdzk | @tmdzk @mattymo @119Vik @rdefosse | Y |
| lte/gateway/ | approvers-agw | @ardzoht @pshelar | @themarwhal @ardzoht @ssanadhya @ulaskozat @pshelar @uri200 @koolzz | Y |
| lte/gateway/python | approvers-agw-python | @ardzoht @pshelar | @ardzoht @pshelar | N |
| lte/gateway/c/session_manager | approvers-agw-sessiond | @themarwhal @uri200 | @themarwhal @uri200 | N |
| lte/gateway/c/oai | approvers-agw-mme | @ssanadhya @ulaskozat @lionelgo @ardzoht | @ssanadhya @ulaskozat @lionelgo @ardzoht | N |
| lte/gateway/c/sctpd | approvers-agw-mme | @ssanadhya @ulaskozat | @ssanadhya @ulaskozat @lionelgo @ardzoht | Y |
| lte/gateway/c/connection_tracker | @ koolzz | @koolzz | @koolzz | N |
| lte/gateway/docker | approvers-ci | @rdefosse @tmdzk | @tmdzk @mattymo @119Vik @rdefosse | Y |
| lte/gateway/release | @tmdzk | @tmdzk | @tmdzk | N |
| lte/gateway/python/magma/pipelined | approvers-agw-pipelined | @koolzz @pshelar | @koolzz @pshelar | N |
| lte/gateway/python/integ_tests | approvers-agw-integ-tests | @ssanadhya @ulaskozat @ardzoht | @ssanadhya @ulaskozat @ardzoht | N |
| lte/gateway/Vagrantfile | approvers-ci | @mattymo @119Vik @tmdzk | @tmdzk @mattymo @119Vik @rdefosse | Y |
| **/*.pb.go | entire-org | entire-org | entire-org | N |
| **/*_swaggergen.go | entire-org | entire-org | entire-org | N |
| xwf/ | approvers-xwf | @AyliD @aharonnovo @r-i-g | @AyliD @aharonnovo @r-i-g | N |
| feg/radius/ | approvers-xwf approvers-feg | @AyliD @aharonnovo @r-i-g @themarwhal @mpgermano @uri200 @emakeev | @AyliD @aharonnovo @r-i-g @themarwhal @mpgermano @uri200 @emakeev | N |
| nms/ | approvers-nms | @karthiksubraveti @andreilee @HannaFar | @karthiksubraveti @andreilee @HannaFar | N |
| CODEOWNERS | @amarpad @electronjoe | @amarpad @electronjoe | @amarpad @electronjoe | N |

**Comment by ulaskozat** [comment]
@themarwhal let's add @ardzoht to `approvers-agw-integ-tests`

**Reply by themarwhal** [reply]
modified doc

**Comment by ekowtaylor** [comment]
@themarwhal planning to discuss this during the eng. community meeting?

**Reply by themarwhal** [reply]
I already discussed briefly on we were considering migrating to this model. It might be worth making a short announcement that we'll try to migrate in the next 1-2 weeks.

**Comment by ekowtaylor** [comment]
What's the workflow for when a PR crosses multiple parts of the codebase?

**Reply by themarwhal** [reply]
Hey @ekowtaylor !
If, for example, my PR touches both feg and orc8r, GitHub will assign the review to `approvers-feg` and `approvers-orc8r` teams. Depending on the team's PR review configuration, each team will assign X (1 by default) team member(s) as reviewers. 

With the current approach, my PR would request reviews from *ALL* orc8r reviewers and *ALL* feg reviewers. 

I've outlined what it looks like from a contributor's perspective in [these slides](https://docs.google.com/presentation/d/1D8DBCczojIQIOyaXwg4VRPNHf1SIN34nc3x8B-fbzdI/edit#slide=id.gce8deb3f59_0_522).

**Comment by koolzz** [comment]
[nit] koolz->koolzz, 

```lte/gateway/c/connection_tracker | @koolz | @koolz | @koolz```

**Reply by themarwhal** [reply]
fixed :)

**Comment by hcgatewood** [comment]
@themarwhal this looks great. See https://github.com/magma/magma/discussions/6301 for my proposal on our GH teams in general, and how the `approvers-*` teams can fit as a nested team within the to-be-created `repo-magma-maintain` team
